### DISCUSSION TITLE
ERC20: transfer amount exceeds balance in fuzz tests

### DISCUSSION THREAD

**Discussion by Gere321123** [Cyfrin/foundry-full-course-cu#2684] [General]
Hi 
In the fuzz test section I get this error:

  [112035] Handler::depositCollateral(73038416682898050331891838846 [7.303e28], 303285569342775139075384605386196978967789599748 [3.032e47])
    ├─ [0] console::log("Bound result", 42186768723657026824961831278 [4.218e28]) [staticcall]
    │   └─ ← [Stop] 
    ├─ [0] VM::prank(0xdD772B89ccC3d513eD1AB23d9356153e8b3438b9)
    │   └─ ← [Return] 
    ├─ [29524] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::mint(0xdD772B89ccC3d513eD1AB23d9356153e8b3438b9, 42186768723657026824961831278 [4.218e28])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0xdD772B89ccC3d513eD1AB23d9356153e8b3438b9, value: 42186768723657026824961831278 [4.218e28])
    │   └─ ← [Stop] 
    ├─ [24647] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::approve(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 42186768723657026824961831278 [4.218e28])
    │   ├─ emit Approval(owner: Handler: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], spender: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 42186768723657026824961831278 [4.218e28])
    │   └─ ← [Return] true
    ├─ [37796] DSCEngine::depositCollateral(0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 42186768723657026824961831278 [4.218e28])
    │   ├─ emit CollateralDeposited(user: Handler: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], token: 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, amount: 42186768723657026824961831278 [4.218e28])
    │   ├─ [5515] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::transferFrom(Handler: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 42186768723657026824961831278 [4.218e28])
    │   │   ├─ emit Approval(owner: Handler: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], spender: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 0)
    │   │   └─ ← [Revert] revert: ERC20: transfer amount exceeds balance
    │   └─ ← [Revert] revert: ERC20: transfer amount exceeds balance
    └─ ← [Revert] revert: ERC20: transfer amount exceeds balance

    
Can somebody help me what is wrong in my code?

foundry.toml file:

    [invariant]
    runs = 128
    depth = 128
    fail_on_revert = true

InvariantsTest.t.sol

     // SPDX-License-Identifier: SEE LICENSE IN LICENSE
    pragma solidity ^0.8.18;
    
    import {Test, console} from "forge-std/Test.sol";
    import {StdInvariant} from "forge-std/StdInvariant.sol";
    import {DeployDSC} from "../../script/DeployDSC.s.sol";
    import {DSCEngine} from "../../src/DSCEngine.sol";
    import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
    import {HelperConfig} from "../../script/HelperConfig.s.sol";
    import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
    import {Handler} from "./Handler.t.sol";
    
    contract InvariantsTest is StdInvariant, Test {
        DeployDSC deployer;
        DSCEngine engine;
        DecentralizedStableCoin dsc;
        HelperConfig helper;
        address weth;
        address wbtc;
        Handler handler;
    
        function setUp() external {
            deployer = new DeployDSC();
            (dsc, engine, helper) = deployer.run();
            (,, weth, wbtc,) = helper.activeNetworkConfig();
            handler = new Handler(engine, dsc);
            targetContract(address(handler));
        }
    
        function invariant_protocolMustHaveMoreValueThanTotalSupply() public view {
            uint256 totalSupply = dsc.totalSupply();
            uint256 totalWethDeposited = IERC20(weth).balanceOf(address(engine));
            uint256 totalBtcDeposited = IERC20(wbtc).balanceOf(address(engine));
    
            uint256 wethValue = engine.getUsdValue(weth, totalWethDeposited);
            uint256 wbtcValue = engine.getUsdValue(wbtc, totalBtcDeposited);
    
            console.log("weth value: ", wethValue);
            console.log("Times mint called: ", handler.timesMintIsCalled());
            assert(wethValue + wbtcValue >= totalSupply);
        }
    
        function invariant_gettersShouldNotRevert() public view {
            engine.getLiquidationBonus();
            engine.getPrecision();
        }
    }

Handler.t.sol

      // SPDX-License-Identifier: SEE LICENSE IN LICENSE
      pragma solidity ^0.8.18;
      
      import {Test, console} from "forge-std/Test.sol";
      import {DSCEngine} from "../../src/DSCEngine.sol";
      import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
      import {ERC20Mock} from "../mocks/ERC20Mock.sol";
      
      contract Handler is Test {
          DSCEngine engine;
          DecentralizedStableCoin dsc;
      
          ERC20Mock weth;
          ERC20Mock wbtc;
      
          uint256 public timesMintIsCalled;
          address[] public userWithCollateralDeposited;
      
          uint256 MAX_DEPOSIT_SIZE = type(uint96).max;
      
          constructor(DSCEngine _dscEngine, DecentralizedStableCoin _dsc) {
              engine = _dscEngine;
              dsc = _dsc;
      
              address[] memory collateralTokens = engine.getCollateralTokens();
              weth = ERC20Mock(collateralTokens[0]);
              wbtc = ERC20Mock(collateralTokens[1]);
          }
      
          function depositCollateral(uint256 collateralSeed, uint256 amountCollateral) public {
              ERC20Mock collateral = _getCollateralFromSeed(collateralSeed);
              amountCollateral = bound(amountCollateral, 1, MAX_DEPOSIT_SIZE);
      
              vm.prank(msg.sender);
              collateral.mint(msg.sender, amountCollateral);
              collateral.approve(address(engine), amountCollateral);
              engine.depositCollateral(address(collateral), amountCollateral);
              vm.stopPrank();
      
              userWithCollateralDeposited.push(msg.sender);
          }
      
          function mintDsc(uint256 amount, uint256 addressSeed) public {
              if (userWithCollateralDeposited.length == 0) {
                  return;
              }
              address sender = userWithCollateralDeposited[addressSeed % userWithCollateralDeposited.length];
              (uint256 totalDscMinted, uint256 collateralValueInUsd) = engine.getAccountInformation(sender);
              if ((collateralValueInUsd / 2) < totalDscMinted) {
                  return;
              }
              uint256 maxDscToMint = ((collateralValueInUsd / 2) - totalDscMinted);
              amount = bound(amount, 0, uint256(maxDscToMint));
              if (amount == 0) {
                  return;
              }
              vm.startPrank(sender);
              engine.mintDsc(amount);
              vm.stopPrank();
      
              timesMintIsCalled++;
          }
      
          function redeemCollateral(uint256 collateralSeed, uint256 amountCollateral) public {
              ERC20Mock collateral = _getCollateralFromSeed(collateralSeed);
              //uint256 maxCollateralToRedeem = engine.getCollateralBalanceOfUser(address(collateral), msg.sender);
              amountCollateral = bound(amountCollateral, 0, MAX_DEPOSIT_SIZE);
              if (amountCollateral == 0) {
                  return;
              }
              engine.redeemCollateral(address(collateral), amountCollateral);
          }
      
          function _getCollateralFromSeed(uint256 collateralSeed) private view returns (ERC20Mock) {
              if (collateralSeed % 2 == 0) {
                  return weth;
              }
              return wbtc;
          }
      }

Many thanks :D

**Comment by EngrPips** [comment]
Hello @Gere321123, Please use `vm.startPrank` instead of `vm.prank`, and let's see how things play out.

**Reply by Gere321123** [reply]
Hi!
yep that was one error tanks, but I still get this error :(

**Reply by EngrPips** [reply]
You will always get errors when writing tests, It is a matter of taking the time to go through the stack trace of the test and read through your test and see where you got it wrong. That is what it means to be a programmer.
